Spigot-Server-Patches/0356-Improve-death-events.patch

```diff
@@ -0,0 +1,279 @@
+ 
+                 if (this.getHealth() <= 0.0F) {
+                     if (!this.e(damagesource)) {
++                        // Paper start - cancellable death event
++                        /*
+                         SoundEffect soundeffect = this.cf();
+ 
+                         if (flag1 && soundeffect != null) {
+                             this.a(soundeffect, this.cq(), this.cr());
+                         }
++                        */
++                        this.silentDeath = !flag1; // mark entity as dying silently
++                        // Paper end
+ 
+                         this.die(damagesource);
++                        this.silentDeath = false; // Paper - cancellable death event - reset to default
+                     }
+                 } else if (flag1) {
+                     this.c(damagesource);
+@@ -1114,6 +1121,8 @@ public abstract class EntityLiving extends Entity {
+             Entity entity = damagesource.getEntity();
+             EntityLiving entityliving = this.ci();
+ 
++            // Paper start - move down to make death event cancellable
++            /*
+             if (this.bb >= 0 && entityliving != null) {
+                 entityliving.a(this, this.bb, damagesource);
+             }
+@@ -1124,6 +1133,9 @@ public abstract class EntityLiving extends Entity {
+ 
+             this.aU = true;
+             this.getCombatTracker().g();
++            */
++            org.bukkit.event.entity.EntityDeathEvent deathEvent = null;
++            //Paper end
+             if (!this.world.isClientSide) {
+                 int i = 0;
+ 
+@@ -1136,15 +1148,32 @@ public abstract class EntityLiving extends Entity {
+ 
+                     this.a(flag, i, damagesource);
+                     // CraftBukkit start - Call death event
+-                    CraftEventFactory.callEntityDeathEvent(this, this.drops);
++                    deathEvent = CraftEventFactory.callEntityDeathEvent(this, this.drops); // Paper - cancellable death event
+                     this.drops = new ArrayList<org.bukkit.inventory.ItemStack>();
+                 } else {
+-                    CraftEventFactory.callEntityDeathEvent(this);
++                    deathEvent = CraftEventFactory.callEntityDeathEvent(this); // Paper - cancellable death event
+                     // CraftBukkit end
+                 }
+             }
+ 
+-            this.world.broadcastEntityEffect(this, (byte) 3);
++            // Paper start - cancellable death event
++            if (deathEvent == null || !deathEvent.isCancelled()) {
++                // triggers and stats got moved down
++                if (this.bb >= 0 && entityliving != null) {
++                    entityliving.a(this, this.bb, damagesource);
++                }
++
++                if (entity != null) {
++                    entity.b(this);
++                }
++
++                this.getCombatTracker().g();
++                this.aU = true; // OBFHELPER - isDying
```

Comment:
yes, single line comments, and that style your referencing is PAIL style from Spigot. They do that to mark 'rename this in the future'.
We want all multi line additions to be unbobfuscated so it's clear what the intended method/property to access/modify is.

With this this.aU, if the isDying field changes, it doesn't tell us where that boolean is in the class / inheritance chain. but if it conflicts on the actual property line, it's very clear, and then every where that property is used by paper is fixed.

Same for that .g() above it too, since its 'copied' logic, needs to be clear what the intent is.

however, I mention single line comments so that if mojang adds or removes lines in this body of code, we can detect that. /* style comments would open us up to desync.

Reply:
Yeah, makes sense, I'll change it to that,